File: Vocoder.cpp
#include <Vocoder.h>

Vocoder::Vocoder(const size_t framesize, const size_t hopsize, const float samplerate) :
  framesize(framesize),
  hopsize(hopsize),
  stft_freq_inc((double)samplerate / (double)framesize),
  stft_phase_inc(PI2 / ((double)framesize / (double)hopsize))
{
  encode_phase_buffer.resize(framesize / 2 + 1);
  decode_phase_buffer.resize(framesize / 2 + 1);
}

void Vocoder::encode(std::vector<std::complex<float>>& dft)
{
  double magnitude,
         frequency,
         phase,
         delta,
         j;

  for (size_t i = 0; i < dft.size(); ++i)
  {
    magnitude = std::abs(dft[i]);
    phase = std::arg(dft[i]);

    delta = phase - encode_phase_buffer[i];
    encode_phase_buffer[i] = phase;

    j = wrap(delta - i * stft_phase_inc) / stft_phase_inc;

    frequency = (i + j) * stft_freq_inc;

    dft[i].real(magnitude);
    dft[i].imag(frequency);
  }
}

void Vocoder::decode(std::vector<std::complex<float>>& dft)
{
  double magnitude,
         frequency,
         phase,
         delta,
         j;

  for (size_t i = 0; i < dft.size(); ++i)
  {
    magnitude = dft[i].real();
    frequency = dft[i].imag();

    j = (frequency - i * stft_freq_inc) / stft_freq_inc;

    delta = (i + j) * stft_phase_inc;

    decode_phase_buffer[i] += delta;
    phase = decode_phase_buffer[i];

    dft[i] = std::polar(magnitude, phase);
  }
}

File: Vocoder.py
import numpy as np


def resample(x, factor):

    N = len(x)
    M = round(N * factor)

    y = np.interp(
        np.linspace(0, 1, M, endpoint=False),
        np.linspace(0, 1, N, endpoint=False),
        x)

    y.resize(x.shape)

    return y


def wrap(x):

    return (x + np.pi) % (2 * np.pi) - np.pi


def encode(frames, framesize, hopsize, sr):

    M, N = frames.shape

    phaseinc = 2 * np.pi * hopsize / framesize
    freqinc = sr / framesize

    buffer = np.zeros(N)
    data = np.zeros((M, N), complex)

    for m, frame in enumerate(frames):

        abs = np.abs(frame)
        arg = np.angle(frame)

        delta = arg - buffer
        buffer = arg

        i = np.arange(N)
        j = wrap(delta - i * phaseinc) / phaseinc

        freq = (i + j) * freqinc

        data[m] = abs + 1j * freq

    return data


def decode(frames, framesize, hopsize, sr):

    M, N = frames.shape

    phaseinc = 2 * np.pi * hopsize / framesize
    freqinc = sr / framesize

    buffer = np.zeros(N)
    data = np.zeros((M, N), complex)

    for m, frame in enumerate(frames):

        abs = np.real(frame)
        freq = np.imag(frame)

        i = np.arange(N)
        j = (freq - i * freqinc) / freqinc

        delta = (i + j) * phaseinc

        buffer += delta
        arg = buffer

        data[m] = abs * np.exp(1j * arg)

    return data
